staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -280,45 +345,106 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 	for _, d := range obj.(Deltas) {
 		switch d.Type {
 		case Sync, Added, Updated:
+			isSync := d.Type == Sync
 			s.cacheMutationDetector.AddObject(d.Object)
 			if old, exists, err := s.indexer.Get(d.Object); err == nil && exists {
 				if err := s.indexer.Update(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object})
+				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object}, isSync)
 			} else {
 				if err := s.indexer.Add(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(addNotification{newObj: d.Object})
+				s.processor.distribute(addNotification{newObj: d.Object}, isSync)
 			}
 		case Deleted:
 			if err := s.indexer.Delete(d.Object); err != nil {
 				return err
 			}
-			s.processor.distribute(deleteNotification{oldObj: d.Object})
+			s.processor.distribute(deleteNotification{oldObj: d.Object}, false)
 		}
 	}
 	return nil
 }
 
 type sharedProcessor struct {
-	listeners []*processorListener
+	listenersLock sync.RWMutex
+	listeners     []*processorListener
+	clock         clock.Clock
+}
+
+func (p *sharedProcessor) addListener(listener *processorListener) {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	p.listeners = append(p.listeners, listener)
 }
 
-func (p *sharedProcessor) distribute(obj interface{}) {
+func (p *sharedProcessor) distribute(obj interface{}, isSync bool) {
+	p.listenersLock.RLock()
+	defer p.listenersLock.RUnlock()
+
 	for _, listener := range p.listeners {
+		if isSync && !listener.shouldHandleSyncItem() {
+			// Skip if the listener doesn't want to resync now
+			continue
+		}
 		listener.add(obj)
 	}
 }
 
+// resyncRunning is called at the start and end of a resync. It passes the
+// event to all the listeners.
+func (p *sharedProcessor) resyncRunning(running bool) {
+	p.listenersLock.RLock()
+	defer p.listenersLock.RUnlock()
+
+	for _, listener := range p.listeners {
+		listener.resyncRunning(running)
```

Comment:
> (the listenersLock, though, is to guard against races adding to the listeners slice)

Pretty sure that changing the value of something in one thread that another thing is reading even if its a member of the slice will be a data race.  Thread 2 may never observe the change, which would break this logic.